dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/DefaultProgramRuleEntityMapperService.java

```diff
@@ -373,4 +390,17 @@ private String getTrackedEntityDataValue( TrackedEntityDataValue dataValue )
 
         return dataValue.getValue() != null ? dataValue.getValue() : "";
     }
+
+    private ValueType getValueTypeForDataElement( String dataElementUid ) {
+        return dataElementToValueTypeCache.get( dataElementUid, () -> {
+            DataElement dataElement = dataElementService.getDataElement( dataElementUid );
```

Comment:
Cache?

Reply:
There is already a cache, for each dataElementUid the query is called only once and then the ValueType of the dataElement is cached with dataElementUid as a key.